watcher/watcher.go

```diff
@@ -472,171 +509,221 @@ func (w *Watcher) WithRedisLock(l *logrus.Entry, f func() error) {
+	})
+
+	if err != nil {
+		l.WithError(err).Error("error listing rooms with no occupied timeout")
+	}
+
+	if len(roomsOnOccupiedTimeout) > 0 {
+		l.WithFields(logrus.Fields{
+			"quantity": len(roomsOnOccupiedTimeout),
+		}).Info("replacing rooms that expired occupied timeout")
+
+		err = w.forwardRemovalRoomEvent(logger, roomsOnOccupiedTimeout)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return roomsOnOccupiedTimeout, nil
+}
+
+func (w *Watcher) forwardRemovalRoomEvent(logger *logrus.Entry, rooms []string) (err error) {
+	// get metadata
+	metadatas, err := models.GetRoomsMetadatas(
+		w.RedisClient.Client,
+		w.SchedulerName, rooms)
+	if err != nil {
+		logger.WithError(err).Error("failed to get rooms metadata")
+		return err
+	}
+
+	// forward
+	for _, roomName := range rooms {
+		room := &models.Room{
+			ID:            roomName,
+			SchedulerName: w.SchedulerName,
+		}
+
+		_, err := eventforwarder.ForwardRoomEvent(
+			context.Background(),
+			w.EventForwarders, w.RedisClient.Client, w.DB, w.KubernetesClient,
+			room, models.RoomTerminated, eventforwarder.PingTimeoutEvent,
+			metadatas[roomName], nil, w.Logger, w.RoomAddrGetter)
+		if err != nil {
+			logger.WithError(err).Error("pingTimeout event forward failed")
+		}
+	}
+	return nil
+}
+
+func (w *Watcher) listPodsToReplace(logger *logrus.Entry, podNames []string) (podsToReplace []v1.Pod, err error) {
+	k := w.KubernetesClient
+	pods, err := k.CoreV1().Pods(w.SchedulerName).List(
+		metav1.ListOptions{})
+	if err != nil {
+		logger.WithError(err).Error("failed to list pods on namespace")
+		return nil, err
+	}
+
+	podNameMap := map[string]bool{}
+	for _, name := range podNames {
+		podNameMap[name] = true
+	}
+
+	for _, pod := range pods.Items {
+		if podNameMap[pod.GetName()] {
+			podsToReplace = append(podsToReplace, pod)
 		}
+	}
 
-		logger.WithFields(logrus.Fields{
-			"rooms": fmt.Sprintf("%v", roomsNoPingSince),
-		}).Info("rooms that are not pinging")
+	return podsToReplace, nil
+}
+
+// RemoveDeadRooms remove rooms that have not sent ping requests for a while
+func (w *Watcher) RemoveDeadRooms() error {
+	w.gracefulShutdown.wg.Add(1)
+	defer w.gracefulShutdown.wg.Done()
+
+	logger := w.Logger.WithFields(logrus.Fields{
+		"executionID": uuid.NewV4().String(),
+		"operation":   "watcher.RemoveDeadRooms",
+	})
+
+	// get rooms with no ping
+	roomsNoPingSince, err := w.roomsWithNoPing(logger)
+	if err != nil {
+		return err
+	}
 
+	// get rooms with occupation timeout
+	roomsOnOccupiedTimeout, err := w.roomsWithOccupationTimeout(logger)
+	if err != nil {
+		return err
+	}
+
+	if len(roomsNoPingSince) > 0 || len(roomsOnOccupiedTimeout) > 0 {
+		podsToReplace, err := w.listPodsToReplace(logger, append(roomsNoPingSince, roomsOnOccupiedTimeout...))
+
+		// load scheduler from database
 		scheduler := models.NewScheduler(w.SchedulerName, "", "")
 		err = scheduler.Load(w.DB)
 		if err != nil {
 			logger.WithError(err).Error("error accessing db while removing dead rooms")
```

Comment:
Right after logging the error, all RemoveDeadRooms does is `logger.Info("finish check of dead rooms")` then it returns nil. Shouldn't you just return this error, remove the else and assume you'll only get there if err == nil?